File: 2021/day_7.js
// https://adventofcode.com/2021/day/7

const assert = require('assert');

const INPUT = require('fs')
  .readFileSync(__dirname + '/day_7.txt', 'utf8')
  .trimEnd()
  .split(',')
  .map(x => parseInt(x));

function calc_median_unsorted(values) {
  const arr = values.slice().sort((a, b) => a - b);
  const mid = Math.floor(arr.length / 2);
  return arr.length % 2 !== 0 ? arr[mid] : (arr[mid - 1] + arr[mid]) / 2;
}

function calc_mean(arr) {
  let sum = 0;
  for (let i = 0; i < arr.length; sum += arr[i++]) {}
  return sum / arr.length;
}

(function part_1() {
  const median = Math.floor(calc_median_unsorted(INPUT));
  let min_fuel = 0;
  for (const x of INPUT) {
    const dist = x < median ? median - x : x - median;
    min_fuel += dist;
  }
  assert.equal(min_fuel, 359648);
})();

(function part_2() {
  const mean = Math.floor(calc_mean(INPUT));
  let min_fuel = 0;
  for (const x of INPUT) {
    const dist = x < mean ? mean - x : x - mean;
    const cost = (dist * (dist + 1)) / 2; // https://en.wikipedia.org/wiki/Summation
    min_fuel += cost;
  }
  assert.equal(min_fuel, 100727924);
})();

File: 2021/day_7.rs
// https://adventofcode.com/2021/day/7

fn main() {
    let mut input: Vec<usize> = include_str!("day_7.txt")
        .trim_end_matches('\n')
        .split(',')
        .map(|x| x.parse::<usize>().unwrap())
        .collect();

    input.sort();

    // Part 1
    {
        let median = input[input.len() / 2];
        let min_fuel = input
            .iter()
            .map(|x| if *x < median { median - x } else { x - median })
            .sum::<usize>();
        assert_eq!(min_fuel, 359648);
    }

    // Part 2
    {
        let avg = input.iter().sum::<usize>() / input.len();
        let min_fuel = input
            .iter()
            .map(|x| if *x < avg { avg - x } else { x - avg })
            .map(|x| {
                (x.pow(2) + x) / 2 // https://en.wikipedia.org/wiki/Summation
            })
            .sum::<usize>();
        assert_eq!(min_fuel, 100727924);
    }
}
